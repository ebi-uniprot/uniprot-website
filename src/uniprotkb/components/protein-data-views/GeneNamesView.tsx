import { FC, Fragment } from 'react';
import { InfoList, ExpandableList } from 'franklin-sites';

import { NameWithEvidence } from './ProteinNamesView';

import { GeneNamesData } from '../../adapters/namesAndTaxonomyConverter';

import { ValueWithEvidence } from '../../types/modelTypes';

const geneAlternativeNamesView = (
  alternativeNames: ValueWithEvidence[],
  firstComma?: boolean
) => (
  <>
    {firstComma && ', '}
    {alternativeNames.map((altName, index) => (
      // eslint-disable-next-line react/no-array-index-key
      <Fragment key={index}>
        {index ? ', ' : undefined}
        <NameWithEvidence data={altName} key={altName.value} />
      </Fragment>
    ))}
  </>
);

const GeneNamesView: FC<{
  geneNamesData: GeneNamesData;
  isCompact?: boolean;
  noTitles?: boolean;
}> = ({ geneNamesData, isCompact = false, noTitles = false }) => (
  <ExpandableList descriptionString="gene names">
    {geneNamesData.map((geneNames, index) => {
      const infoData = [
        {
          title: 'Name',
          content: geneNames.geneName && (
            <NameWithEvidence data={geneNames.geneName} />
          ),
        },
      ];
      if (geneNames.synonyms) {
        infoData.push({
          title: 'Synonyms',
          content: geneAlternativeNamesView(geneNames.synonyms),
        });
      }
      if (geneNames.orfNames) {
        infoData.push({
          title: 'ORF names',
          content: geneAlternativeNamesView(geneNames.orfNames),
        });
      }
      if (geneNames.orderedLocusNames) {
        infoData.push({
          title: 'Ordered locus names',
          content: geneAlternativeNamesView(geneNames.orderedLocusNames),
        });
      }
      return (
        <InfoList
          infoData={infoData}
          key={geneNames.geneName?.value || index}
          isCompact={isCompact}
          highlightFirstItem={isCompact}
          noTitles={noTitles}
        />
      );
    })}
  </ExpandableList>
);

export default GeneNamesView;
