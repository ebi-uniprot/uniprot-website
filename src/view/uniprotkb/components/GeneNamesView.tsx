import React, { Fragment } from 'react';
import { v1 } from 'uuid';
import { InfoList } from 'franklin-sites';
import { GeneNamesData } from '../../../model/uniprotkb/sections/NamesAndTaxonomyConverter';

export const geneAlternativeNamesView = (
  alternativeNames: { value: string }[],
  firstComma: boolean = true
) => {
  const altNames = alternativeNames.map(altName => altName.value).join(', ');
  return `${firstComma ? ', ' : ''}${altNames}`;
};

const GeneNamesView: React.FC<{
  geneNamesData: GeneNamesData;
  isCompact?: boolean;
}> = ({ geneNamesData, isCompact = false }) => (
  <Fragment>
    {geneNamesData.map(geneNames => {
      const infoData = [
        {
          title: 'Name',
          content: geneNames.geneName && (
            <Fragment>{geneNames.geneName.value}</Fragment>
          ),
        },
      ];
      if (geneNames.synonyms) {
        infoData.push({
          title: 'Synonyms',
          content: (
            <Fragment>
              {geneAlternativeNamesView(geneNames.synonyms, false)}
            </Fragment>
          ),
        });
      }
      if (geneNames.orfNames) {
        infoData.push({
          title: 'ORF names',
          content: (
            <Fragment>
              {geneAlternativeNamesView(geneNames.orfNames, false)}
            </Fragment>
          ),
        });
      }
      if (geneNames.orderedLocusNames) {
        infoData.push({
          title: 'Ordered locus names',
          content: (
            <Fragment>
              {geneAlternativeNamesView(geneNames.orderedLocusNames, false)}
            </Fragment>
          ),
        });
      }
      return (
        <InfoList
          infoData={infoData}
          key={geneNames.geneName ? geneNames.geneName.value : v1()}
          isCompact={isCompact}
          highlightFirstItem={isCompact}
        />
      );
    })}
  </Fragment>
);

export default GeneNamesView;
