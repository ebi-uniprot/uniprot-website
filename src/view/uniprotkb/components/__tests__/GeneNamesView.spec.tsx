import React from 'react';
import { render } from '@testing-library/react';
import GeneNamesView from '../GeneNamesView';
import GeneNamesUIDataJson from '../__mocks__/GeneNamesUIData.json';

describe('GeneNames', () => {
  test('should render gene_names', () => {
    const { asFragment } = render(<GeneNamesView {...GeneNamesUIDataJson} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
