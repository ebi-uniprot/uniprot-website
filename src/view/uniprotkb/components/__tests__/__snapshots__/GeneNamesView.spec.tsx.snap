// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneNames should render gene_names 1`] = `
<DocumentFragment>
  <div
    class="info-list"
  >
    <div
      class="info-list__item"
    >
      <div
        class="info-list__item__title"
      >
        Name
      </div>
      <div
        class="info-list__item__content"
      >
        some Gene
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__item__title"
      >
        Synonyms
      </div>
      <div
        class="info-list__item__content"
      >
        some Syn
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__item__title"
      >
        ORF names
      </div>
      <div
        class="info-list__item__content"
      >
        some orf
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__item__title"
      >
        Ordered locus names
      </div>
      <div
        class="info-list__item__content"
      >
        some locus
      </div>
    </div>
  </div>
</DocumentFragment>
`;
